pkg/pmem-csi-driver/nodeserver.go

```diff
@@ -97,47 +98,74 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 	defer volumeMutex.UnlockKey(req.GetVolumeId())
 
 	targetPath := req.TargetPath
-	stagingtargetPath := req.StagingTargetPath
-	// TODO: check is bind-mount already made
-	// (happens when publish is asked repeatedly for already published namespace)
-	notMnt, err := mount.New("").IsLikelyNotMountPoint(targetPath)
-	if err != nil && !os.IsNotExist(err) {
-		return nil, status.Error(codes.Internal, "validate target path: "+err.Error())
-	}
-	if !notMnt {
-		// TODO(https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/95): check if mount is compatible. Return OK if it is, or appropriate error.
-		/*
-			1) Target Path MUST be the vol referenced by vol ID
-			2) VolumeCapability MUST match
-			3) Readonly MUST match
-		*/
-		return &csi.NodePublishVolumeResponse{}, nil
-	}
-
-	if err := os.Mkdir(targetPath, os.FileMode(0755)); err != nil {
-		// Kubernetes is violating the CSI spec and creates the
-		// directory for us
-		// (https://github.com/kubernetes/kubernetes/issues/75535). We
-		// allow that by ignoring the "already exists" error.
-		if !os.IsExist(err) {
-			return nil, status.Error(codes.Internal, "make target dir: "+err.Error())
+	sourcePath := req.StagingTargetPath
+	mounter := mount.New("")
+
+	switch req.VolumeCapability.GetAccessType().(type) {
+	case *csi.VolumeCapability_Block:
+		dev, err := ns.dm.GetDevice(req.VolumeId)
+		if err != nil {
+			return nil, status.Error(codes.NotFound, "No device found with volume id "+req.VolumeId+": "+err.Error())
+		}
+		// For block volumes, source path is the actual Device path
+		sourcePath = dev.Path
+		targetDir := filepath.Dir(targetPath)
+		// Make sure that parent directory of target path is existing, otherwise create it
+		targetPreExisting, err := ensureDirectory(mounter, targetDir)
+		if err != nil {
+			return nil, err
+		}
+		err = mounter.MakeFile(targetPath)
+		if err != nil && !os.IsExist(err) {
+			if !targetPreExisting {
+				if rerr := os.Remove(targetDir); rerr != nil {
+					glog.Warningf("Could not remove created mount target %q: %v", targetDir, rerr)
+				}
+			}
+			return nil, status.Errorf(codes.Internal, "Could not create target device file %q: %v", targetPath, err)
+		}
+	case *csi.VolumeCapability_Mount:
+		// TODO: check is bind-mount already made
+		// (happens when publish is asked repeatedly for already published namespace)
+		notMnt, err := mounter.IsLikelyNotMountPoint(targetPath)
```

Comment:
Here we use  IsLikelyNotMountPoint, I guess thats another one I picked in early stage of this project. But comment says, although it's faster, this method is not 100% reliable. So what about changing this to IsNotMountPoint which should be reliable. We don't have run-time pressure so we should value reliability over performance.

Reply:
Agree, will do that change

Reply:
Done.